app/usecases/leaderboards.py

```diff
@@ -23,14 +28,47 @@ async def create(beatmap: Beatmap, mode: Mode) -> Leaderboard:
         leaderboard.scores.append(score)
 
     leaderboard.sort()
+    await insert_cache(beatmap, leaderboard)
     return leaderboard
 
 
+# We expect to handle locking on a per-case basis.
 async def fetch(beatmap: Beatmap, mode: Mode) -> Leaderboard:
-    if leaderboard := beatmap.leaderboards.get(mode):
+    if leaderboard := await fetch_cache(beatmap, mode):
         return leaderboard
 
     leaderboard = await create(beatmap, mode)
-    beatmap.leaderboards[mode] = leaderboard
 
     return leaderboard
+
+
+async def fetch_cache(beatmap: Beatmap, mode: Mode) -> Optional[Leaderboard]:
+
+    redis_cache = await app.state.services.redis.get(
+        f"ussr:leaderboards:{beatmap.md5}:{mode.value}",
+    )
+
+    if not redis_cache:
+        return None
+
+    await app.state.services.redis.expire(
+        f"ussr:leaderboards:{beatmap.md5}:{mode.value}",
+        timedelta(days=1),
+    )
+
+    score_dicts = json.loads(redis_cache)
+
+    return Leaderboard(
+        mode=mode,
+        scores=[Score.from_dict(score_dict) for score_dict in score_dicts],
+    )
+
+
+# Assumes the lock has already been acquired.
+async def insert_cache(beatmap: Beatmap, leaderboard: Leaderboard) -> None:
```

Comment:
i would name this generically: `insert`. usecases should be universal and generally have specific names avoided

Reply:
Fixed